addons/mail/wizard/mail_compose_message.py

```diff
@@ -992,6 +992,45 @@ def _prepare_mail_values_dynamic(self, res_ids):
                     reply_to = mail_values.get('email_from', False)
                 mail_values['reply_to'] = reply_to
 
+            # body: render layout in email mode (comment mode is managed by the
+            # notification process, see @_notify_thread_by_email)
+            if email_mode and self.email_layout_xmlid and mail_values['recipient_ids']:
+                lang = langs[res_id]
+                recipient_ids = [command[1] for command in mail_values['recipient_ids']]
+                record_wlang = record.with_context(lang=lang)
+
+                msg_vals = {
+                    'email_layout_xmlid': self.email_layout_xmlid,
+                    'model': record_wlang._name,
+                    'res_id': record_wlang.id,
+                }
+                message_inmem = self.env['mail.message'].new({
+                    'body': mail_values['body'],
+                })
+
+                for _lang, render_values, recipients_group_data in record_wlang._notify_by_email_iterate_recipients(
+                    message_inmem,
+                    [{
+                        'active': True,
+                        'id': pid,
+                        'lang': False,
+                        'notif': 'email',
+                        'share': True,
+                        'type': 'customer',
+                        'ushare': False,
+                    } for pid in recipient_ids],
+                    msg_vals=msg_vals,
+                    model_description=False,  # force dynamic computation
+                    force_email_lang=lang,
+                ):
+                    mail_body = record_wlang._notify_by_email_render_layout(
+                        message_inmem,
+                        recipients_group_data,
+                        msg_vals=msg_vals,
+                        render_values=render_values,
+                    )
+                    mail_values['body_html'] = mail_body
```

Comment:
Does this mean that the body is rendered once through "render_field" and then we override it here? Or do we just encapsulate the body into a proper template with the "message_inmem" thing?

Does that mean that it was not encapsulated before?
So we add multiple features here actually?

Maybe commit message is a bit light then.

Reply:
Indeed it is a complete new feature. I forgot to update the commit message, but currently if you use the composer in mass mailing mode, it just sends the body. If you use the composer to post a message, emails are sent with a layout. Here we now support using a layout in mailing mode, to try to make both modes more coherent.